sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/WithFields.scala

```diff
@@ -17,16 +17,29 @@
 
 package org.apache.spark.sql.catalyst.optimizer
 
-import org.apache.spark.sql.catalyst.expressions.WithFields
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, GetStructField, WithFields}
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.catalyst.rules.Rule
 
 
 /**
- * Combines all adjacent [[WithFields]] expression into a single [[WithFields]] expression.
+ * Optimizes [[WithFields]] expression chains.
  */
-object CombineWithFields extends Rule[LogicalPlan] {
+object OptimizeWithFields extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+    case WithFields(structExpr, names, values) if names.distinct.length != names.length =>
+      val newNames = mutable.ArrayBuffer.empty[String]
+      val newValues = mutable.ArrayBuffer.empty[Expression]
+      names.zip(values).reverse.foreach { case (name, value) =>
+        if (!newNames.contains(name)) {
+          newNames += name
+          newValues += value
+        }
+      }
+      WithFields(structExpr, names = newNames.reverse.toSeq, valExprs = newValues.reverse.toSeq)
```

Comment:
You are right. It is eventually the same. This is coming from improving scalability of #29587. This is applied during I fixed the scalability issue. I found this is useful to reduce the complex of `WithFields` expression tree.

I will run these rules in #29587 to simplify expression tree before optimizer.

Reply:
Okay, so I took a look at the PR you linked and left a related [comment](https://github.com/apache/spark/pull/29587/files#r493098043) there. I don't think you actually need this optimization for `unionByName`. 

This optimization is only useful if someone uses `WithFields` to update the same field multiple times. However, it would simply be better to **not** update the same field multiple times. At the very least, we should not do this when we re-use this Expression internally within Spark. 

Unfortunately, "bad" end-users might still update the same field multiple times. Assuming we should optimize for such users (not sure), since this batch is only applied half-way through the optimization cycle, I think we could just move up the `Batch("ReplaceWithFieldsExpression", Once, ReplaceWithFieldsExpression)` to get the same benefit (which is just simplified tree). What do you reckon?

Reply:
Actually I'd like to run these rules to simplify `WithFields` tree early in analysis stage. After #29587, I thought that it is very likely to write bad `WithFields` tree. Once hitting that, it is very hard to debug and the analyzer/optimizer spend a lot of time traversing expression tree. So I think it is very useful keep this rule to simplify the expression tree, but I don't think we want to do `ReplaceWithFieldsExpression` in analysis stage.

Reply:
ahh I see, yes, in the analysis stage this would likely be helpful!

Okay in that case, could this PR wait till #29795 goes in? I'm refactoring `WithFields` so this optimization would need to change anyway.

Reply:
I'm fine to wait until #29795.